File: hashes/2131-longest-palindrome-by-concatenating-two-letter-words.py
from collections import Counter
from typing import List
import unittest

# https://leetcode.com/problems/longest-palindrome-by-concatenating-two-letter-words/

# python3 -m unittest hashes/2131-longest-palindrome-by-concatenating-two-letter-words.py


class Solution(unittest.TestCase):
    def longestPalindrome(self, words: List[str]) -> int:
        counter = Counter(words)
        length, center = 0, 0
        for word, cnt in counter.items():
            revw = word[::-1]
            if word == revw:
                length += cnt // 2 * 4
                if cnt & 1:
                    center = 2
            else:
                length += min(cnt, counter[revw]) * 2
        return length + center

    def test(self):
        for words, expected in [
            (["lc", "cl", "gg"], 6),
            (["cc", "ll", "xx"], 2),
            (["lc", "cl", "gg", "ak", "kk"], 6),
            (["ab", "ty", "yt", "lc", "cl", "ab"], 8),
            (["ab", "ty", "yt", "lc", "cl", "ab", "ba"], 12),
            (["dd", "aa", "bb", "dd", "aa", "dd", "bb", "dd", "aa", "cc", "bb", "cc", "dd", "cc"], 22),
            (["qo", "fo", "fq", "qf", "fo", "ff", "qq", "qf", "of", "of", "oo", "of", "of", "qf", "qf", "of"], 14),
        ]:
            output = self.longestPalindrome(words)
            self.assertEqual(expected, output, f"expected: {expected}, output: {output}")

File: hashes/2131-longest-palindrome-by-concatenating-two-letter-words.go
package hashes

// https://leetcode.com/problems/longest-palindrome-by-concatenating-two-letter-words/

// Approach #2: Array
func longestPalindrome(words []string) int {
	var count [26 * 26]int
	length, unpaired := 0, 0
	var a, b, word, revw int
	for idx := range words {
		a = int(words[idx][0] - 'a')
		b = int(words[idx][1] - 'a')
		word, revw = a*26+b, b*26+a

		if count[revw] > 0 {
			length += 4
			count[revw]--
			if word == revw {
				unpaired--
			}
		} else {
			count[word]++
			if word == revw {
				unpaired++
			}
		}
	}
	if unpaired > 0 {
		length += 2 // center
	}
	return length
}

// // Approach #1: Hash Table
// func longestPalindrome(words []string) int {
// 	var length = 0
// 	var count = make(map[[2]byte]int)
// 	var word, revw [2]byte
// 	for idx := range words {
// 		word[0], word[1] = words[idx][0], words[idx][1]
// 		revw[0], revw[1] = words[idx][1], words[idx][0]
// 		if count[revw] > 0 {
// 			count[revw]--
// 			length += 4
// 		} else {
// 			count[word]++
// 		}
// 	}
// 	for word, cnt := range count {
// 		if cnt > 0 && word[0] == word[1] {
// 			length += 2
// 			break
// 		}
// 	}
// 	return length
// }
